### ISSUE TITLE
BABE: epochs

### ISSUE THREAD

**Issue by rphmeier** [paritytech/substrate#3032]
BABE has a notion of epoch, which is a fixed number of slots where the validator set is the same. Within the context of the SRML, an epoch is the same as a session.

Every epoch has its own randomness, which is fully-determined 1 epoch ahead of time but may be different on forks longer than 1 epoch.
After #2946 and #2970 , the next session validators are also fully-determined 1 session ahead of time.

A block author proves that it wins a slot by computing the slot number and epoch randomness and proving that it falls below some threshold which is dynamically computed from their proportion of the total stake-weight.

-----

For import, it's nice to have the ability to do initial verification of many blocks without having to execute them one-by-one. This is definitely possible for all blocks within the same epoch (by which I mean all blocks with a common epoch-ancestor, so they have the same epoch randomness). 

For this we'll want to store all "relevant" epoch-heads on the node-side as well as data about the upcoming epoch. By relevant I mean unfinalized or within the finalized chain. Anything else can be pruned.

What we probably want to do is have BABE issue signals whenever an epoch ends:
```rust
struct BabeEpochSignal {
    next_epoch_validators: Vec<(PublicKey, u64)>, 
    next_epoch_randomness: [u8; 32],
}
```

Then the node-side code knows whenever it sees a block with this signal at slot number S, the first descendent of this block in any chain with slot number >= `S + SLOTS_PER_EPOCH`, will trigger an epoch change to one with `next_epoch_validators` and `next_epoch_randomness`. Because it is fully-determined 1 epoch in advance, this should be fairly safe for light clients except in the case of long network partitions.

**Comment by Kianenigma** [comment]
I assumed linked + assigned to https://github.com/paritytech/substrate/pull/3028?

**Comment by rphmeier** [comment]
Yes, that should cover a good chunk of it
